pandas/core/reshape/concat.py

```diff
@@ -584,7 +595,17 @@ def get_result(self):
                 cons = sample._constructor_expanddim
 
                 index, columns = self.new_axes
-                df = cons(data, index=index, copy=self.copy)
+                mgr = dict_to_mgr(
+                    data,
+                    index,
+                    None,
+                    copy=self.copy,
+                    typ=get_option("mode.data_manager"),
+                )
+                if using_copy_on_write() and not self.copy:
+                    mgr.parent = [obj._mgr for obj in self.objs]
+                    mgr.refs = [weakref.ref(obj._mgr.blocks[0]) for obj in self.objs]
```

Comment:
I would very much prefer that we can keep this ref handling only to the internals .. 
I was thinking, in theory it should be possible to let `dict_to_mgr` do this? (that might actually also help for the `DataFrame(dict of Series)` constructor case)

Reply:
(that might be quite a bit more complicated though (since that function needs to handle more than just this case), so we could also leave that for a follow-up)

Reply:
I agree in general, but we have to be very careful. 

Would you be ok with pushing this to the internals after finishing the constructors or when doing the constructors? We don't have many tests in the constructor area yet and I don't want to break anything without noticing.

Reply:
>  pushing this to the internals after finishing the constructors or when doing the constructors? 

I am doing this in https://github.com/pandas-dev/pandas/pull/50777. Will update that PR to also refactor the above once this is merged